File: ConversationIDE/src/VanEngine.Core/VAN/fryas_alphabet.py
from __future__ import annotations

from .juul_mask import JuulMask


class FryasAlphabet:
    _map: dict[str, JuulMask] = {
        # Vowels (7)
        "I": JuulMask.SPOKE0,
        "—": JuulMask.SPOKE180,
        "O": JuulMask.OUTER_RIM,
        "Λ": JuulMask.SPOKE0 | JuulMask.SPOKE60,
        "V": JuulMask.SPOKE120 | JuulMask.SPOKE180,
        "/": JuulMask.SPOKE60,
        "\\": JuulMask.SPOKE300,
        # Core Consonants (16)
        "M": JuulMask.SPOKE0 | JuulMask.SPOKE180,
        "N": JuulMask.SPOKE60 | JuulMask.SPOKE240,
        "T": JuulMask.SPOKE0 | JuulMask.SPOKE60 | JuulMask.SPOKE120,
        "K": JuulMask.SPOKE180 | JuulMask.SPOKE240 | JuulMask.SPOKE300,
        "F": JuulMask.SPOKE0 | JuulMask.SPOKE180 | JuulMask.OUTER_RIM,
        "S": JuulMask.SPOKE0 | JuulMask.SPOKE60 | JuulMask.SPOKE120 | JuulMask.SPOKE180 | JuulMask.SPOKE240 | JuulMask.SPOKE300,
        "R": JuulMask.SPOKE60 | JuulMask.SPOKE180 | JuulMask.SPOKE300,
        "H": JuulMask.SPOKE240 | JuulMask.SPOKE300 | JuulMask.SPOKE0,
        # Remaining 8 consonants
        "B": JuulMask.SPOKE0 | JuulMask.SPOKE60 | JuulMask.OUTER_RIM,
        "D": JuulMask.SPOKE60 | JuulMask.SPOKE180 | JuulMask.OUTER_RIM,
        "G": JuulMask.SPOKE120 | JuulMask.SPOKE240 | JuulMask.SPOKE300,
        "L": JuulMask.SPOKE0 | JuulMask.SPOKE300,
        "P": JuulMask.SPOKE0 | JuulMask.SPOKE60 | JuulMask.SPOKE180,
        "W": JuulMask.SPOKE60 | JuulMask.SPOKE120 | JuulMask.SPOKE240,
        "J": JuulMask.SPOKE120 | JuulMask.SPOKE300,
        "Z": JuulMask.SPOKE0 | JuulMask.SPOKE240 | JuulMask.SPOKE300 | JuulMask.OUTER_RIM,
        # Numerals (5) — all include OUTER_RIM to distinguish from vowels
        "0": JuulMask.OUTER_RIM | JuulMask.SPOKE0 | JuulMask.SPOKE120,
        "1": JuulMask.OUTER_RIM | JuulMask.SPOKE60 | JuulMask.SPOKE300,
        "2": JuulMask.OUTER_RIM | JuulMask.SPOKE0 | JuulMask.SPOKE240,
        "3": JuulMask.OUTER_RIM | JuulMask.SPOKE120 | JuulMask.SPOKE300,
        "4": JuulMask.OUTER_RIM | JuulMask.SPOKE60 | JuulMask.SPOKE180 | JuulMask.SPOKE240,
        # Extended marks (6)
        "Þ": JuulMask.SPOKE0 | JuulMask.SPOKE60 | JuulMask.SPOKE240,
        "Ð": JuulMask.SPOKE60 | JuulMask.SPOKE180 | JuulMask.SPOKE240,
        "Æ": JuulMask.SPOKE0 | JuulMask.SPOKE120 | JuulMask.SPOKE180,
        "Œ": JuulMask.SPOKE60 | JuulMask.SPOKE120 | JuulMask.SPOKE240 | JuulMask.OUTER_RIM,
        "×": JuulMask.SPOKE0 | JuulMask.SPOKE120 | JuulMask.SPOKE240,
        "†": JuulMask.SPOKE60 | JuulMask.SPOKE180 | JuulMask.SPOKE300 | JuulMask.OUTER_RIM,
    }

    @staticmethod
    def get_mask(character: str) -> JuulMask:
        mask = FryasAlphabet._map.get(character)
        if mask is None:
            raise KeyError(f"Character '{character}' is not in the Fryas alphabet.")
        return mask

    @staticmethod
    def get_character(mask: JuulMask) -> str | None:
        for ch, m in FryasAlphabet._map.items():
            if m == mask:
                return ch
        return None

File: engine/VanEngine.Core/VAN/FryasAlphabet.cs
namespace VanEngine.Core.VAN;

public static class FryasAlphabet
{
    public static IReadOnlyDictionary<char, JuulMask> Map => _map;
    private static readonly Dictionary<char, JuulMask> _map = new()
    {
        // ── Vowels (7) ──────────────────────────────────────────
        { 'I', JuulMask.Spoke0 },
        { '—', JuulMask.Spoke180 },
        { 'O', JuulMask.OuterRim },
        { 'Λ', JuulMask.Spoke0 | JuulMask.Spoke60 },
        { 'V', JuulMask.Spoke120 | JuulMask.Spoke180 },
        { '/', JuulMask.Spoke60 },
        { '\\', JuulMask.Spoke300 },

        // ── Core Consonants (16) ────────────────────────────────
        { 'M', JuulMask.Spoke0 | JuulMask.Spoke180 },
        { 'N', JuulMask.Spoke60 | JuulMask.Spoke240 },
        { 'T', JuulMask.Spoke0 | JuulMask.Spoke60 | JuulMask.Spoke120 },
        { 'K', JuulMask.Spoke180 | JuulMask.Spoke240 | JuulMask.Spoke300 },
        { 'F', JuulMask.Spoke0 | JuulMask.Spoke180 | JuulMask.OuterRim },
        { 'S', JuulMask.Spoke0 | JuulMask.Spoke60 | JuulMask.Spoke120 | JuulMask.Spoke180 | JuulMask.Spoke240 | JuulMask.Spoke300 },
        { 'R', JuulMask.Spoke60 | JuulMask.Spoke180 | JuulMask.Spoke300 },
        { 'H', JuulMask.Spoke240 | JuulMask.Spoke300 | JuulMask.Spoke0 },

        // Remaining 8 consonants — awaiting Captain's Oera Linda chart
        { 'B', JuulMask.Spoke0 | JuulMask.Spoke60 | JuulMask.OuterRim },
        { 'D', JuulMask.Spoke60 | JuulMask.Spoke180 | JuulMask.OuterRim },
        { 'G', JuulMask.Spoke120 | JuulMask.Spoke240 | JuulMask.Spoke300 },
        { 'L', JuulMask.Spoke0 | JuulMask.Spoke300 },
        { 'P', JuulMask.Spoke0 | JuulMask.Spoke60 | JuulMask.Spoke180 },
        { 'W', JuulMask.Spoke60 | JuulMask.Spoke120 | JuulMask.Spoke240 },
        { 'J', JuulMask.Spoke120 | JuulMask.Spoke300 },
        { 'Z', JuulMask.Spoke0 | JuulMask.Spoke240 | JuulMask.Spoke300 | JuulMask.OuterRim },

        // ── Numerals (5) — all include OuterRim to distinguish from vowels ──
        { '0', JuulMask.OuterRim | JuulMask.Spoke0 | JuulMask.Spoke120 },
        { '1', JuulMask.OuterRim | JuulMask.Spoke60 | JuulMask.Spoke300 },
        { '2', JuulMask.OuterRim | JuulMask.Spoke0 | JuulMask.Spoke240 },
        { '3', JuulMask.OuterRim | JuulMask.Spoke120 | JuulMask.Spoke300 },
        { '4', JuulMask.OuterRim | JuulMask.Spoke60 | JuulMask.Spoke180 | JuulMask.Spoke240 },

        // ── Extended marks (6) — awaiting Captain's chart ──
        { 'Þ', JuulMask.Spoke0 | JuulMask.Spoke60 | JuulMask.Spoke240 },
        { 'Ð', JuulMask.Spoke60 | JuulMask.Spoke180 | JuulMask.Spoke240 },
        { 'Æ', JuulMask.Spoke0 | JuulMask.Spoke120 | JuulMask.Spoke180 },
        { 'Œ', JuulMask.Spoke60 | JuulMask.Spoke120 | JuulMask.Spoke240 | JuulMask.OuterRim },
        { '×', JuulMask.Spoke0 | JuulMask.Spoke120 | JuulMask.Spoke240 },
        { '†', JuulMask.Spoke60 | JuulMask.Spoke180 | JuulMask.Spoke300 | JuulMask.OuterRim },

        // Total: 7 + 16 + 5 + 6 = 34
    };

    public static JuulMask GetMask(char character)
    {
        if (_map.TryGetValue(character, out var mask))
            return mask;
        throw new ArgumentOutOfRangeException(nameof(character), $"Character '{character}' is not in the Fryas alphabet.");
    }

    public static char? GetCharacter(JuulMask mask)
    {
        foreach (var kv in _map)
            if (kv.Value == mask)
                return kv.Key;
        return null;
    }
}
